### PR TITLE
MCMC diagnostics

### PR DESCRIPTION
**Pull request by AlexManochis**
This PR implements in C++:

1. Geweke MCMC diagnostic test
2. Gelman-Rubin Potential Scale Reduction Factor (PSRF)
3. Raftery & Lewis MCMC diagnostic test

- Geweke is implemented as Hotelling's Two Sample test, which is a multivariate extension of the common 
two sample Student's t-test.   
- Gelman-Rubin multivariate case given in https://www.tandfonline.com/doi/pdf/10.1080/10618600.1998.10474787?needAccess=true is implemented.  
- Implementation of Raftery diagnostic is based on Matlab version of coda package in http://www.spatial-econometrics.com/gibbs/  

It provides also Rcpp functions to export those implementations in R.

### CODE REVIEW

File: include/diagnostics/geweke.hpp [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+//Licensed under GNU LGPL.3, see LICENCE file
+
+
+#ifndef GEWEKE_HPP
+#define GEWEKE_HPP
+
+#include <boost/math/distributions/fisher_f.hpp>
+
+template <typename VT, typename MT, typename NT>
+bool perform_geweke(MT const& samples, NT const& frac1, NT const& frac2)
```
Comment (vissarion) [inline include/diagnostics/geweke.hpp]: please provide some reference of the method (as you did for the R interface) and a comment on what is `frac1`, `frac2`.
Should `frac1`, `frac2` have default values? Also is there any other better name, e.g. `frac_first`, `frac_last` ?

Comment (vissarion) [reply include/diagnostics/geweke.hpp]: What is this function doing exactly? What is the meaning of the returned bool ?

File: include/diagnostics/psrf.hpp [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+//Contributed and/or modified by Alexandros Manochis, as part of Google Summer of Code 2020 program.
+
+//Licensed under GNU LGPL.3, see LICENCE file
+
+
+#ifndef PSRF_HPP
+#define PSRF_HPP
+
+template <typename NT, typename VT, typename MT>
+NT perform_psrf(MT const& points)
```
Comment (vissarion) [inline include/diagnostics/psrf.hpp]: here there are not frac1, frac2 and the return type is NT compared to bool above

File: include/diagnostics/raftery.hpp [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+#include "raftery_subroutines/empquant.hpp"
+#include "raftery_subroutines/indtest.hpp"
+#include "raftery_subroutines/mctest.hpp"
+#include "raftery_subroutines/mcest.hpp"
+#include "raftery_subroutines/thin.hpp"
+#include "raftery_subroutines/ppnd.hpp"
+
+
+template <typename VT, typename MT, typename NT>
+std::pair<MT,VT> perform_raftery(MT const& samples, NT const& q, NT const& r, NT const& s)
```
Comment (vissarion) [inline include/diagnostics/raftery.hpp]: another different interface: returns a pair and there are some q,r,s parameters that is unclear to understand

**Review by vissarion — APPROVED** [review]
I am ok in general, looks good! Are you going to add tests/examples?

File: include/diagnostics/geweke.hpp (hunk 1 above)
Comment (AlexManochis) [reply include/diagnostics/geweke.hpp]: I have added comments to provide information about geweke diagnostic

File: include/diagnostics/psrf.hpp (hunk 2 above)
Comment (AlexManochis) [reply include/diagnostics/psrf.hpp]: I have added comments to provide information about psrf diagnostic

File: include/diagnostics/raftery.hpp (hunk 3 above)
Comment (AlexManochis) [reply include/diagnostics/raftery.hpp]: I have added comments to provide information about raftery diagnostic

**Review by vissarion — APPROVED** [review]
Thanks for the revision.

**Review by TolisChal — APPROVED** [review]

**Pull request merged**